### DISCUSSION TITLE
DeDRM Conversion fails : Mac OS 11,6,1, Kindle 1,33, DeDRM 7,2,1

### DISCUSSION THREAD

**Discussion by HDMW1** [apprenticeharper/DeDRM_tools#1943] [Q&A]
Hello :
I copy the log file below along with the requested information.
I've read the FAQs and tried the suggested trouble shooting.

Error : "The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required." 
 I don't know how to locate this file.

Regards : Hubert Wagner

Mac OS 11.6.1
eBook from Amazon
Kindle v 1,33 (620000). Kindle for Mac 44173
DeDRM tool V 7.2.1 & v6.8.0 and v6.8.1
All results for KindleForMac-44182.dmg lead to an Apple page which cannot be reached when download is chosen.

Terminal
chmod -x /Applications/Kindle.app/Contents/MacOS/renderer-test
No such file or directory

calibre Debug log
calibre 5.32  embedded-python: True is64bit: True
macOS-10.16-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '20.6.0', 'Darwin Kernel Version 20.6.0: Tue Oct 12 18:33:38 PDT 2021; root:xnu-7195.141.8~1/RELEASE_ARM64_T8101')
Python 3.8.5
OSX: ('10.16', ('', '', ''), 'x86_64')
Interface language: en_GB
Successfully initialized third party plugins: DeDRM (7, 2, 1)
calibre 5.32  embedded-python: True is64bit: True
macOS-10.16-x86_64-i386-64bit Darwin ('64bit', '')
('Darwin', '20.6.0', 'Darwin Kernel Version 20.6.0: Tue Oct 12 18:33:38 PDT 2021; root:xnu-7195.141.8~1/RELEASE_ARM64_T8101')
Python 3.8.5
OSX: ('10.16', ('', '', ''), 'x86_64')
Interface language: en_GB
Successfully initialized third party plugins: DeDRM (7, 2, 1)
devicePixelRatio: 2.0
logicalDpi: 72.0 x 72.0
physicalDpi: 149.00000223775552 x 149.0000022377555
Using calibre Qt style: True
[0.00] Starting up...
[0.01] Showing splash screen...
[0.22] splash screen shown
[0.22] Initializing db...
[0.26] db initialized
[0.26] Constructing main UI...
[0.83] main UI initialized...
[0.83] Hiding splash screen
Starting QuickView
DeDRM v7.2.1: Trying to decrypt B09GVWYWYK_EBOK.azw
Using Library AlfCrypto Python
Using Library AlfCrypto Python
DeDRM v7.2.1: Failed to decrypt with error: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.
DeDRM v7.2.1: Looking for new default Kindle Key after 0.1 seconds
Found k4Mac kinf2018 file: /Users/hubertwagner/Library/Application Support/Kindle/storage/.kinf2018
trying username  b'hubertwagner'  on file  /Users/hubertwagner/Library/Application Support/Kindle/storage/.kinf2018
trying IDString: b'bba5f6fabfcd'
trying IDString: b'bba5f6fabfcc'
trying IDString: b'151b2653a4e3'
[11.97] splash screen hidden
[11.97] Started up in 11.97 seconds with 1674 books
Decrypted key file using IDString '151b2653a4e3' and UserName 'hubertwagner'
DeDRM v7.2.1: Found 1 new key
DeDRM v7.2.1: Ultimately failed to decrypt after 1.2 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Running file type plugin DeDRM failed with traceback:
Traceback (most recent call last):
  File "calibre_plugins.dedrm.__init__", line 540, in KindleMobiDecrypt
    book = k4mobidedrm.GetDecryptedBook(path_to_ebook,kindleDatabases,androidFiles,serials,pids,self.starttime)
  File "calibre_plugins.dedrm.k4mobidedrm", line 208, in GetDecryptedBook
    raise DrmException("The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.")
calibre_plugins.dedrm.k4mobidedrm.DrmException: The .kfx DRMION file cannot be decrypted by itself. A .kfx-zip archive containing a DRM voucher is required.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "calibre/customize/ui.py", line 175, in _run_filetype_plugins
  File "calibre_plugins.dedrm.__init__", line 633, in run
    decrypted_ebook = self.KindleMobiDecrypt(path_to_ebook)
  File "calibre_plugins.dedrm.__init__", line 583, in KindleMobiDecrypt
    raise DeDRMError("{0} v{1}: Ultimately failed to decrypt after {2:.1f} seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md".format(PLUGIN_NAME, PLUGIN_VERSION,time.time()-self.starttime))
calibre_plugins.dedrm.DeDRMError: DeDRM v7.2.1: Ultimately failed to decrypt after 1.2 seconds. Read the FAQs at Harper's repository: https://github.com/apprenticeharper/DeDRM_tools/blob/master/FAQs.md
Traceback (most recent call last):
  File "calibre/customize/ui.py", line 432, in get_file_type_metadata
  File "calibre/customize/builtins.py", line 260, in get_metadata
  File "calibre/ebooks/metadata/mobi.py", line 504, in get_metadata
  File "calibre/ebooks/mobi/reader/headers.py", line 294, in __init__
  File "calibre/ebooks/mobi/reader/headers.py", line 322, in identity
calibre.ebooks.mobi.MobiError: Unknown book type: b'\x81\xc4\xde\x02\x9d\x8c#\x01'

**Comment by ElleKayEm** [comment]
If you want to decrypt a KFX format book, you need to leave the files in their original folder structure.  But KFX format downloaded with Kindle for Mac later than version 1.26 cannot be decrypted.  Alternatively, disabling KFX format downloads to get the older format requires version 1.31 or earlier.  Take a look at this post:  https://www.mobileread.com/forums/showpost.php?p=3819708&postcount=508

**Comment by j-howell** [comment]
It is probably best to avoid KFX format, however if you do want to import KFX into calibre you need to have the KFX Input plugin installed for it to work properly.

**Comment by HDMW1** [comment]
Hello Mr Harper :
Thank you for you almost instant response !

If Calibre 5.x and tools 7.2.1+ aren't working, try getting Calibre 4.23 and tools 6.8.1 working first, then install 5.x & 7.2.1+ on top of that.

This worked first time.

Best Wishes : Hubert Wagner, France

**Comment by HDMW1** [comment]
Hello Mr Howell :
Thank you for your remarks.

> On 24 Nov 2021, at 14:37, John Howell ***@***.***> wrote:
> It is probably best to avoid KFX format, however if you do want to import KFX into calibre you need to have the KFX Input plugin installed for it to work properly.


Here’s a screen shot of a recent download from Amazon. This appears to be KFX format.
KFX plugin 1,45 installed in calibre v4,23. 

I use the French Cybook Muse by Bookeen so every eBook I buy needs to be converted.
Everything is working smoothly thanks to the work of the Apprentice group.



Regards : Hubert Wagner, France

**Comment by HDMW1** [comment]
One last note.
I downloaded the eBook with v 1,33 of Kindle for Mac.
Converted with DeDRM tool 6,8,1 - However, when loaded into calibre the tool appears as 6,8,0
calibre v4,2,3 - The KFX plugin was already installed (unless this came from calibre saved preferences - I installed the plugin in the latest version I had been using before down shifting to v4,2,3

I thank the community for their generous responses.
Best Wishes : Hubert Wagner, France

**Comment by ElleKayEm** [comment]
Can you confirm what version of Kindle for Mac you have?Other Mac users have reported that it is not possible to disable KFX downloads from Kindle for Mac 1.33 (or 1.32).  And decryption of KFX format only works with books downloaded from Kindle for Mac 1.26 or earlier.  The book in the earlier log was KFX.

**Comment by HDMW1** [comment]
Hello :
Thank you for your assistance.

My Kindle for Mac is v 1.33.0 62000, Last updated 11/7/2021
Note also Mac OS 11.6.1 M1 chip

I downloaded the book first. I didn’t open it with Kindle.

Regards : Hubert Wagner

**Comment by ElleKayEm** [comment]
So you were able to disable KFX downloads from Kindle for Mac 1.33?

**Comment by HDMW1** [comment]
Hello  :
The short answer is ‘yes’.

Kindle-for-Mac is 1,33
Note : Kindle-for-Mac is running in Rosetta on the M1 Mac.
DeDRM tool is 6.8.1 although calibre show 6.8.0 after installation.
callibre v4.23.0

The book is in calibre and opens with eBook viewer.
And I’ve sent it to my Cybook Muse where it opens correctly.

If you come across anything more of interest, please let me know.

Thank you again for your interest : Hubert Wagner

**Comment by ElleKayEm** [comment]
That's good news because others had reported it no longer worked after 1.31.

You should be able to update back to the current calibre and DeDRM now that the key is found.

**Comment by HDMW1** [comment]
Hello :

> On 24 Nov 2021, at 19:25, ElleKayEm ***@***.***> wrote:
> That's good news because others had reported it no longer worked after 1.31.
> 
> You should be able to update back to the current calibre and DeDRM now that the key is found.

I don’t understand your remark about the ‘key’.

Please look at the original calibre log. 
https://github.com/apprenticeharper/DeDRM_tools/discussions/1943

The conversion did not work with Kindle 1.33.0, the current version of calibre 5.x and DeDRM 7.2.1
I assume this would still be the case with future downloads from Amazon. No?

Regards : Hubert Wagner

**Reply by ElleKayEm** [reply]
The problem with Macs and DeDRM 7.2.1 (calibre 5.x) is only the initial retrieval of the key from Kindle for Mac.  After you've successfully decrypted a book with calibre 4.23 and DeDRM 6.8.1, you should be able to update.

**Comment by HDMW1** [comment]
> On 24 Nov 2021, at 19:48, ElleKayEm ***@***.***> wrote:
> 
> The problem with Macs and DeDRM 7.2.1 (calibre 5.x) is only the initial retrieval of the key from Kindle for Mac. After you've successfully decrypted a book with calibre 4.23 and DeDRM 6.8.1, you should be able to update.

Hello  :
Does this mean the ‘key’ is a serial number which after the initial conversion is aftwards kept in calibre preferences?
And that key is available to updated version of calibre?

Regards : Hubert Wagner

**Reply by ElleKayEm** [reply]
Each installation of Kindle for Mac has a unique key which DeDRM retrieves and decrypts.  If it successfully decrypts a book with that key, it is saved and used to decrypt any other books downloaded with the same app.  And yes, the key remains if you update calibre and DeDRM.
